packages/drivers/odsp-driver/src/fetchSnapshot.ts

```diff
@@ -239,14 +240,40 @@ async function redeemSharingLink(
 				const encodedShareUrl = getEncodedShareUrl(
 					odspResolvedUrl.shareLinkInfo?.sharingLinkToRedeem,
 				);
-				const redeemUrl = `${odspResolvedUrl.siteUrl}/_api/v2.0/shares/${encodedShareUrl}`;
-				const { url, headers } = getUrlAndHeadersWithAuth(
-					redeemUrl,
-					storageToken,
-					forceAccessTokenViaAuthorizationHeader,
-				);
-				headers.prefer = "redeemSharingLink";
-				return fetchAndParseAsJSONHelper(url, { headers });
+
+				let redeemUrl: string | undefined;
+				async function callSharesAPI(baseUrl: string): Promise<void> {
+					redeemUrl = `${baseUrl}/_api/v2.0/shares/${encodedShareUrl}`;
+					const { url, headers } = getUrlAndHeadersWithAuth(
+						redeemUrl,
+						storageToken,
+						forceAccessTokenViaAuthorizationHeader,
+					);
+					headers.prefer = "redeemSharingLink";
+					await fetchAndParseAsJSONHelper(url, { headers });
+				}
+
+				const disableUsingTenantDomain = loggerToMonitoringContext(
+					logger,
+				).config.getBoolean("Fluid.Driver.Odsp.DisableUsingTenantDomainForSharesApi");
+				// There is an issue where if we use the siteUrl in /shares, then the allowed length of url is just a few hundred characters
+				// and we fail to do the redeem. But if we use the tenant domain in the url, then the allowed length becomes 2048. So, first
+				// construct the url for /shares using tenant domain but to be on safer side, fallback to what we were using before.
+				if (!disableUsingTenantDomain) {
+					try {
+						await callSharesAPI(new URL(odspResolvedUrl.siteUrl).origin);
+						return;
+					} catch (error) {
+						logger.sendTelemetryEvent(
+							{
+								eventName: "ShareLinkRedeemFailedWithDomain",
+								length: redeemUrl?.length,
```

Comment:
Should we add additional params here to help understand when an issue occurs? I don't know what part of the URL we can safely log but we should log what we can.

Reply:
The error message would tell us if it is due to length.

Reply:
Yes, true. I was thinking if that will be enough for customers / applications to debug issues. Should we add additional info to help apps like Loop, outlook, etc. understand what the issue is if this log is seen.